library/cwm/src/lib/cwm/ui_state.rb

```diff
@@ -0,0 +1,209 @@
+    #
+    # @return [String] Textdomain (e.g., "firewall")
+    abstract_method :textdomain_name
+
+    # Method to be called when the user decides to visit a given page by
+    # clicking in one node of the general tree.
+    #
+    # It remembers the decision so the user is taken back to a sensible point of
+    # the tree (very often the last he decided to visit) after redrawing.
+    #
+    # @param [CWM::Page] page associated to the tree node
+    def go_to_tree_node(page)
+      self.candidate_nodes = [page.label]
+
+      # Landing in a new node, so invalidate previous details about position
+      # within a node, they no longer apply
+      self.tab = nil
+    end
+
+    # Method to be called when the user switches to a tab within a tree node.
+    #
+    # It remembers the decision so the same tab is showed in case the user stays
+    # in the same node after redrawing.
+    #
+    # @param [CWM::Page] page associated to the tab
+    def switch_to_tab(page)
+      self.tab = page.label
+    end
+
+    # Method to be called when the user operates in a row of a table.
+    #
+    # @param row_id [Object] row identifier
+    def select_row(row_id)
+      self.row_id = row_id
+    end
+
+    # Select the page to open in the general tree after a redraw
+    #
+    # @param pages [Array<CWM::Page>] all the pages in the tree
+    # @return [CWM::Page, nil]
+    def find_tree_node(pages)
+      candidate_nodes.each.with_index do |candidate, idx|
+        result = pages.find { |page| matches?(page, candidate) }
+        if result
+          # If we had to use one of the fallbacks, the tab name is not longer
+          # trustworthy
+          self.tab = nil unless idx.zero?
+          return result
+        end
+      end
+      self.tab = nil
+      nil
+    end
+
+    # Select the tab to open within the node after a redraw
+    #
+    # @param pages [Array<CWM::Page>] pages for all the possible tabs
+    # @return [CWM::Page, nil]
+    def find_tab(pages)
+      return nil unless tab
+      pages.find { |page| page.label == tab }
+    end
+
+    # @!attribute [r] row_id
+    #   @return [Object] Selected row identifier
+    # @see #row_id
+    attr_reader :row_id
+
+  protected
+
+    attr_writer :row_id
+
+    # Where to place the user within the general tree in next redraw
+    # @return [Array<Integer, String>]
+    attr_accessor :candidate_nodes
+
+    # Concrete tab within the current node to show in the next redraw
+    # @return [String, nil]
+    attr_reader :tab
+    # @see #tab
+    def tab=(tab)
+      @tab = tab
+      # If the user switched to a new tab, invalidate details about the inner
+      # table
+      self.row_id = nil
+    end
+
+    # Whether the given page matches with the candidate tree node
+    #
+    # @param page [CWM::Page]
+    # @param candidate [Integer, String]
+    # @return boolean
+    def matches?(page, candidate)
+      page.label == candidate
+    end
+
+    class << self
```

Comment:
why not use Singleton module?

Reply:
To be able to create a new instance during tests.

Reply:
it is possible even with standard Singleton module:

just try this code snippet

```
jreidinger@linux-vvcf:~/prace/yast/storage-ng> cat /tmp/test.rb 
require "singleton"

class A
  include Singleton

  attr_accessor :text
end

A.instance.text = "test"

puts A.instance.text.inspect
Singleton.__init__(A)
puts A.instance.text.inspect

```

Reply:
To be honest, calling to a method called `__init__` looks wrong to me. I prefer to just add a few lines of code than relying on a method which is not expected to be used outside. It is a common pattern in yast2-storage-ng.